README.md
wsgi2cgi
========

Run CGI apps under Python WSGI protocol (PEP 333).

This is a simple WSGI application that executes an external process
and translates the WSGI protocol to the Common Gateway Interface (CGI).

Example:

    from wsgiref.simple_server import make_server
    import wsgi2cgi

    def app(environ, start_response):
       wapp = wsgi2cgi.CGI('/path/to/executable.cgi')
       return wapp.application(environ, start_response)

    httpd = make_server('127.0.0.1', 8000, app)
    print "Serving on 127.0.0.1:8000..."
    httpd.serve_forever()

The CGI class supports two arguments:

 - command: absolute path to the CGI application executable.
 - extra\_env: additional environment variables. 

The WSGI application can be used with any server supporting WSGI.

More information:

 - PEP-333: http://www.python.org/dev/peps/pep-0333/
 - WSGI Servers: http://wsgi.readthedocs.org/en/latest/servers.html


Install
-------

The module can be installed using setup.py:

    python setup.py install

Alternativelly it can be installed with pip:

    pip install wsgi2cgi


License
-------

This package is open source under the MIT license terms. Check COPYING
file for further details.


Author
------

 - Juan J. Martinez <jjm@usebox.net>



CHANGES
2012-07-18 Juan J. Martinez <jjm@usebox.net>

    * Release 0.2.1
        * Small fixes to look OK in PyPi

2012-04-28 Juan J. Martinez <jjm@usebox.net>

    * Release 0.2
        * Small fixes
        * Tests

2012-04-25 Juan J. Martinez <jjm@usebox.net>

    * Release 0.1
        * Fist public release



tests/get.py
#!/usr/bin/env python

from StringIO import StringIO

from base import BaseTestCase
import wsgi2cgi

CGI_TEST_JSON = """#!/usr/bin/env python
import os
import json

print "Content-Type: text/plain\\n"
print json.dumps(dict(os.environ))
"""

CGI_TEST_JSON_STATUS = """#!/usr/bin/env python
import os
import json

print "Status: %s"
print "Content-Type: text/plain\\n"
print json.dumps(dict(os.environ))
"""

class TestGetMethod(BaseTestCase):

    def test_wsgiref_defaults(self):
        self.cgi.write(CGI_TEST_JSON)
        self.cgi.close()

        response = self.request()

        self.assertEqual(self.status, "200 OK")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertEqual(response['REQUEST_METHOD'], 'GET')

        for key, value in response.iteritems():
            if not key.startswith("HTTP_"):
                self.assertTrue(key in wsgi2cgi.CGI_VARS)

    def test_http_headers(self):
        self.cgi.write(CGI_TEST_JSON)
        self.cgi.close()

        self.env['HTTP_EXTRA_HEADER'] = 'extra-http-header'
        response = self.request()

        self.assertEqual(self.status, "200 OK")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertEqual(response['REQUEST_METHOD'], 'GET')

        for key, value in response.iteritems():
            if not key.startswith("HTTP_"):
                self.assertTrue(key in wsgi2cgi.CGI_VARS)

        self.assertEqual(response['HTTP_EXTRA_HEADER'], 'extra-http-header')

    def test_extra_env(self):
        self.cgi.write(CGI_TEST_JSON)
        self.cgi.close()

        self.extra_env = dict(EXTRA_HEADER='extra-header')
        response = self.request()

        self.assertEqual(self.status, "200 OK")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertEqual(response['REQUEST_METHOD'], 'GET')
        self.assertEqual(response['EXTRA_HEADER'], 'extra-header')

    def test_status_ok(self):
        self.cgi.write(CGI_TEST_JSON_STATUS % "200 OK")
        self.cgi.close()

        response = self.request()

        self.assertEqual(self.status, "200 OK")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertEqual(response['REQUEST_METHOD'], 'GET')

    def test_status_not_found(self):
        self.cgi.write(CGI_TEST_JSON_STATUS % "404 Not Found")
        self.cgi.close()

        response = self.request()

        self.assertEqual(self.status, "404 Not Found")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertEqual(response['REQUEST_METHOD'], 'GET')

    def test_internal_error_popen_format_error(self):
        self.cgi.write("FORMAT-ERROR")
        self.cgi.close()

        self.env['wsgi.errors'] = StringIO()
        response = self.request_raw()

        self.assertEqual(self.status, "500 Internal Server Error")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertTrue("Exec format error" in self.env['wsgi.errors'].getvalue())
        self.env['wsgi.errors'].close()

    def test_internal_error_popen_not_such_file(self):
        self.cgi.write("#!/donotexist\n")
        self.cgi.close()

        self.env['wsgi.errors'] = StringIO()
        response = self.request_raw()

        self.assertEqual(self.status, "500 Internal Server Error")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertTrue("No such file" in self.env['wsgi.errors'].getvalue())
        self.env['wsgi.errors'].close()

    def test_internal_error_invalid_content_length(self):
        self.cgi.write(CGI_TEST_JSON)
        self.cgi.close()

        self.env['CONTENT_LENGTH'] = 'invalid'
        self.env['wsgi.errors'] = StringIO()
        response = self.request_raw()

        self.assertEqual(self.status, "500 Internal Server Error")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertTrue("invalid Content-Length header" in self.env['wsgi.errors'].getvalue())
        self.env['wsgi.errors'].close()

    def test_internal_error_invalid_header(self):
        self.cgi.write('#!/bin/sh\n\necho -e "Content-Type: text/plain\\nInvalid-Header value\n\nERROR"')
        self.cgi.close()

        self.env['wsgi.errors'] = StringIO()
        response = self.request_raw()

        self.assertEqual(self.status, "500 Internal Server Error")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertTrue("invalid header" in self.env['wsgi.errors'].getvalue())
        self.env['wsgi.errors'].close()


if __name__ == '__main__':
        unittest.main()



tests/base.py

import json
import os
import tempfile
import unittest

from wsgiref.util import setup_testing_defaults

import wsgi2cgi

class BaseTestCase(unittest.TestCase):
    def setup_app(self, environ, start_response):
        wapp = wsgi2cgi.CGI(self.cgi.name, extra_env=self.extra_env)
        return wapp.application(environ, start_response)

    def start_response(self, status, headers):
        self.status = status
        self.headers = headers

    def request_raw(self):
        app = self.setup_app(self.env, self.start_response)
        return ''.join(app)

    def request(self):
        return json.loads(self.request_raw())

    def setUp(self):
        self.cgi = tempfile.NamedTemporaryFile(delete=False)
        os.chmod(self.cgi.name, 0700)

        self.env = dict()
        setup_testing_defaults(self.env)

        self.extra_env = None

    def tearDown(self):
        os.unlink(self.cgi.name)



tests/post.py
#!/usr/bin/env python

from StringIO import StringIO

from base import BaseTestCase
import wsgi2cgi

CGI_TEST = """#!/usr/bin/env python
import os
import sys

print "Content-Type: text/plain\\n"
print "METHOD=%s" % os.environ['REQUEST_METHOD']
indata = sys.stdin.read()
print indata
"""

class TestPostMethod(BaseTestCase):

    def test_post(self):
        self.cgi.write(CGI_TEST)
        self.cgi.close()

        expected = "Some content to be post\nsome content for the test"

        self.env['REQUEST_METHOD'] = 'POST'
        self.env['wsgi.input'] = StringIO(expected)
        self.env['CONTENT_LENGTH'] = str(len(expected))

        response = self.request_raw()

        self.assertEqual(self.status, "200 OK")
        self.assertEqual(self.headers, [('Content-Type', 'text/plain')])

        self.assertTrue(response.startswith('METHOD=POST\n'))
        self.assertEqual(response[len('METHOD=POST\n'):], expected + "\n")


if __name__ == '__main__':
        unittest.main()



wsgi2cgi/__init__.py
"""
Run CGI apps under Python WSGI protocol (PEP 333).

This is a simple WSGI application that executes an external process
and translates the WSGI protocol to the Common Gateway Interface (CGI).

For example usage, see the CGI class.
"""
__author__  =  'Juan J. Martinez'
__version__ =  '0.2.1'
__all__ = "CGI"

import sys
import re
from subprocess import Popen, PIPE, STDOUT

# Allowed CGI environmental variables
CGI_VARS = """
SERVER_SOFTWARE
SERVER_NAME
GATEWAY_INTERFACE
SERVER_PROTOCOL
SERVER_PORT
REQUEST_METHOD
PATH_INFO
PATH_TRANSLATED
SCRIPT_NAME
QUERY_STRING
REMOTE_HOST
REMOTE_ADDR
AUTH_TYPE
REMOTE_USER
REMOTE_IDENT
CONTENT_TYPE
CONTENT_LENGTH
""".split('\n')

# Buffer size for buffered input/output
BUFFER = 1024*64

class CGI(object):
    """
    Run a CGI app with WSGI.

    Example:
    >>> from wsgiref.simple_server import make_server
    >>> import wsgi2cgi

    >>> def app(environ, start_response):
    >>>    wapp = wsgi2cgi.CGI('/path/to/executable.cgi')
    >>>    return wapp.application(environ, start_response)

    >>> httpd = make_server('127.0.0.1', 8000, app)
    >>> print "Serving on 127.0.0.1:8000..."
    >>> httpd.serve_forever()

    """
    def __init__(self, command, extra_env=None):
        """
        CGI class constructor.

        Args:
            command: the command to be executed (absolute path).
            extra_env: additional environment variables, useful to pass
                configuration information to some CGI applications.

        Raises:
            ValueError: extra_env is not a dictionary.
        """
        self.cmd = [arg.strip() for arg in command.split(' ')]
        self.extra_env = extra_env
        self.env = dict()

        if extra_env and not isinstance(extra_env, dict):
            raise ValueError("extra_env is not a dictionary")

    def log_error(self, message):
        """
        Logs errors to wsgi.errors.

        Args:
            message: string to be logged.

        The destination of the error messages depends on the WSGI server.
        """
        fd = self.env.get('wsgi.errors', sys.stderr)
        fd.write("%s: %s\n" % (self.cmd, message))
        fd.flush()

    def application(self, environ, start_response):
        """
        WSGI application.

        Args:
            environ: WSGI enviroment.
            start_response: start response callable.

        Any internal error executing the CGI application will be logged
        in wsgi.errors and a HTTP 500 error response will be sent to the
        client.
        """
        self.env = environ
        cgi_env = dict()
        for key, value in environ.iteritems():
            if key in CGI_VARS or key.startswith('HTTP_'):
                cgi_env[key] = value

        if self.extra_env:
            cgi_env.update(self.extra_env)

        try:
            process = Popen(self.cmd, stdin=PIPE, stdout=PIPE, stderr=STDOUT, env=cgi_env)
        except (OSError, ValueError), e:
            self.log_error(str(e))
            start_response("500 Internal Server Error", [('Content-Type', 'text/plain')])
            yield "500 Internal Server Error\n"
            return

        try:
            content_length = int(environ.get('CONTENT_LENGTH', 0) or 0)
        except ValueError:
            self.log_error("invalid Content-Length header")
            start_response("500 Internal Server Error", [('Content-Type', 'text/plain')])
            yield "500 Internal Server Error\n"
            return

        stdin = environ.get('wsgi.input', None)
        while stdin and content_length > 0:
            data = stdin.read(BUFFER)
            if not data:
                break
            content_length -= len(data)
            process.stdin.write(data)
        process.stdin.close()

        response = None
        headers = []
        while True:
            line = process.stdout.readline().strip()
            if not line:
                break

            if re.match(r'Status: [0-9]{3}.*', line):
                if response:
                    self.log_error('duplicated Status header: %s' % line)
                response = line[len("Status :"):]
                continue

            if ':' not in line:
                self.log_error('invalid header: %s' % line)
                start_response("500 Internal Server Error", [('Content-Type', 'text/plain')])
                yield "500 Internal Server Error\n"
                return

            header, value = line.split(':', 1)
            headers.append((header.strip(), value.strip()))

        start_response(response or "200 OK", headers)

        while True:
            data = process.stdout.read(BUFFER)
            if not data:
                break
            yield data

        process.stdout.close()
        return
